snakemake/executors.py

```diff
@@ -2087,3 +2101,658 @@ def run_wrapper(
+                    )
+                )
+            return f
+
+        self.workflow_sources = []
+
+        for wfs in self.workflow.get_sources():
+            if os.path.isdir(wfs):
+                for (dirpath, dirnames, filenames) in os.walk(wfs):
+                    self.workflow_sources.extend(
+                        [check_size(os.path.join(dirpath, f)) for f in filenames]
+                    )
+            else:
+                self.workflow_sources.append(check_size(os.path.abspath(wfs)))
+
+    def _generate_build_source_package(self):
+        """in order for the instance to access the working directory in storage,
+           we need to upload it. This file is cleaned up at the end of the run.
+           We do this, and then obtain from the instance and extract.
+        """
+        # Add the execution scripts
+        # import code
+        # code.interact(local=locals())
+
+        # We will generate a tar.gz package, renamed by hash
+        tmpname = next(tempfile._get_candidate_names())
+        targz = os.path.join(tempfile.gettempdir(), "snakemake-%s.tar.gz" % tmpname)
+        tar = tarfile.open(targz, "w:gz")
+
+        # Add all workflow_sources files
+        for filename in self.workflow_sources:
+            arcname = filename.replace(self.workflow.basedir + os.path.sep, "")
+            tar.add(filename, arcname=arcname)
+
+        # Add snakemake hidden directory
+        for root, dirs, files in os.walk(os.path.join(self.workdir, ".snakemake")):
+            for filename in files:
+                filename = os.path.join(root, filename)
+                if not filename.endswith("lock") and not filename.startswith(
+                    "snakeworkdir"
+                ):
+                    arcname = filename.replace(self.workdir + os.path.sep, "")
+                    tar.add(filename, arcname=arcname)
+
+        tar.close()
+
+        # Rename based on hash, in case user wants to save cache
+        sha256 = get_file_hash(targz)
+        hash_tar = os.path.join(self.workdir, "snakeworkdir-%s.tar.gz" % sha256)
+
+        # Only copy if we don't have it yet, clean up if we do
+        if not os.path.exists(hash_tar):
+            shutil.move(targz, hash_tar)
+        else:
+            os.remove(targz)
+
+        # We will clean these all up at shutdown
+        self._build_packages.add(hash_tar)
+
+        return hash_tar
+
+    def _generate_job_action(self, job):
+        """generate a single action to execute the job.
+        """
+        # Write the jobscript, this will be the container entrypoint along
+        # with obtaining the working directory for the run
+        jobscript = self.get_jobscript(job)
+        self.write_jobscript(job, jobscript)
+
+        # Remote files (within Snakefile) should be referenced via remotes.GS
+        # For the workdir, we need to upload it, named based on hash of contents
+        targz = self._generate_build_source_package()
```

Comment:
Why don't you create the tarball once, upon construction of the executor? Then, we don't need to recreate it for each job.

Reply:
Definitely, let me fix that. I think I originally made this choice thinking that the build packages would be different for each step, but they aren't.

Reply:
Working on this now.

Reply:
okay I have an answer for why we cannot do that. The job file (the entrypoint) is generated on the level of a job step, so if we only copy the first we get an error:

```bash
/bin/bash: .snakemake/tmp.hzxks7k9/snakejob.copy.2.sh: No such file or directory
```
This is why it's done for each job.

Reply:
I don't think you need to create a script as an entrypoint. It should be possible to do what I do in the kubernetes executor (directly pass a shell command instead):
* https://github.com/snakemake/snakemake/blob/master/snakemake/executors.py#L1397-L1402
* https://github.com/snakemake/snakemake/blob/master/snakemake/executors.py#L1417

Reply:
It's generated by snakemake, and includes all the properties / configs, etc. as is done locally. Here is an example:

```bash
#!/bin/sh
# properties = {"type": "single", "rule": "copy", "local": false, "input": ["gcp-public-data-landsat/LC08/01/001/003/LC08_L1GT_001003_20170430_20170501_01_RT/LC08_L1GT_001003_20170430_20170501_01_RT_MTL.txt"], "output": ["snakemake-testing/1/landsat-data.txt"], "wildcards": {}, "params": {}, "log": [], "threads": 1, "resources": {"mem_mb": 100}, "jobid": 2, "cluster": {}}
snakemake snakemake-testing/1/landsat-data.txt --snakefile Snakefile --force -j --keep-target-files  --keep-remote --latency-wait 0 --attempt 1 --force-use-threads  --allowed-rules copy --nocolor --notemp --no-hooks --nolock  --use-conda  --default-remote-provider GS --default-remote-prefix snakemake-testing/1 
```
I don't think I understand how it would work without that.

Reply:
The entrypoint is already overloaded, I don't see how we can add more to it (e.g., see here https://github.com/snakemake/snakemake/pull/90/files#diff-c923148fbfa18c64d8a76ab68afd7b79R2520)

Reply:
Actually I see what you mean, let me give it a try.

Reply:
okay, all fixed, both the entrypoint update (not relying on a script) and consequently, also the generation of the build package (only done once!)

Reply:
okay this is all fixed.